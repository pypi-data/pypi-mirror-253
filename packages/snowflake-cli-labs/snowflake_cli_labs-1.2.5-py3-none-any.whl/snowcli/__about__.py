from __future__ import annotations

VERSION = "1.2.5"
