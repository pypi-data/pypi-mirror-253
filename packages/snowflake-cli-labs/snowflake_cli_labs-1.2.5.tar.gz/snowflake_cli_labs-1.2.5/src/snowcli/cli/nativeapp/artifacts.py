import os
import shutil
from click.exceptions import ClickException
from pathlib import Path
from typing import List, Union
from dataclasses import dataclass


class ArtifactError(ClickException):
    """
    Could not parse source or destination artifact.
    """

    def __init__(self, msg: str):
        super().__init__(msg)


class GlobMatchedNothingError(ClickException):
    """
    No files were found that matched the provided glob pattern.
    """

    def __init__(self, src: str):
        super().__init__(f"{self.__doc__}: {src}")


class SourceNotFoundError(ClickException):
    """
    The specifically-referenced source file or directory was not found
    in the project directory.
    """

    path: Path

    def __init__(self, path: Path):
        super().__init__(f"{self.__doc__}\npath = {path}")
        self.path = path


class TooManyFilesError(ClickException):
    """
    Multiple files were mapped to one output file.
    """

    dest_path: Path

    def __init__(self, dest_path: Path):
        super().__init__(f"self.__doc__\ndest_path = {dest_path}")
        self.dest_path = dest_path


class OutsideDeployRootError(ClickException):
    """
    The specified path is outside of the deploy root.
    This can happen when a relative path with ".." is provided.
    """

    dest_path: Path
    deploy_root: Path

    def __init__(self, dest_path: Path, deploy_root: Path):
        super().__init__(
            f"""
            self.__doc__
            \ndest_path = {dest_path}
            \ndeploy_root = {deploy_root}
            """
        )
        self.dest_path = dest_path
        self.deploy_root = deploy_root


@dataclass
class ArtifactMapping:
    """
    Used to keep track of equivalent paths / globs so we can copy
    artifacts from the project folder to the deploy root.
    """

    src: str
    dest: str


def is_glob(s: str) -> bool:
    return "*" in s


def specifies_directory(s: str) -> bool:
    """
    Does the path (as seen from the project definition) refer to
    a directory? For destination paths, we enforce the usage of a
    trailing slash (i.e. \ for windows; / for others).

    This means that to put a file in the root of the stage, we need
    to specify "./" as its destination, or omit it (but only if the
    file already lives in the project root).
    """
    return s.endswith(os.sep)


def symlink_or_copy(src: Path, dst: Path, makedirs=True) -> None:
    """
    Tries to create a symlink to src at dst; failing that (i.e. in Windows
    without Administrator / Developer Mode) copies the file from src to dst instead.
    If makedirs is True, the directory hierarchy above dst is created if any
    of those directories do not exist.
    """
    if makedirs:
        dst.parent.mkdir(parents=True, exist_ok=True)
    try:
        os.symlink(src, dst)
    except OSError:
        shutil.copy(src, dst)


def translate_artifact(item: Union[dict, str]) -> ArtifactMapping:
    if isinstance(item, dict):
        return ArtifactMapping(item["src"], item.get("dest", item["src"]))

    elif isinstance(item, str):
        return ArtifactMapping(item, item)

    # XXX: validation should have caught this
    raise ArtifactError("Item is not a valid artifact!")


def get_source_paths(artifact: ArtifactMapping, project_root: Path) -> List[Path]:
    """
    Expands globs, ensuring at least one file exists that matches artifact.src.
    Returns a list of paths that resolve to actual files in the project root dir structure.
    """
    source_paths: List[Path]

    if is_glob(artifact.src):
        source_paths = list(project_root.glob(artifact.src))
        if not source_paths:
            raise GlobMatchedNothingError(artifact.src)
    else:
        source_path = Path(project_root, artifact.src)
        source_paths = [source_path]
        if not source_path.exists():
            raise SourceNotFoundError(source_path)

    return source_paths


def build_bundle(
    project_root: Path, deploy_root: Path, artifacts: List[ArtifactMapping]
):
    """
    Prepares a local folder (deploy_root) with configured app artifacts.
    This folder can then be uploaded to a stage.
    """
    resolved_root = deploy_root.resolve()
    if resolved_root.exists() and not resolved_root.is_dir():
        raise ValueError(f"Deploy root {resolved_root} exists, but is not a directory!")

    for artifact in artifacts:
        # make sure we are only modifying files / directories inside the deploy root
        dest_path = Path(resolved_root, artifact.dest).resolve()
        if resolved_root != dest_path and resolved_root not in dest_path.parents:
            raise OutsideDeployRootError(dest_path, resolved_root)

        if dest_path.is_file():
            dest_path.unlink()

        source_paths = get_source_paths(artifact, project_root)

        if specifies_directory(artifact.dest):
            # copy all files as children of the given destination path
            for source_path in source_paths:
                symlink_or_copy(source_path, dest_path / source_path.name)
        else:
            if len(source_paths) == 1:
                # copy a single file as the given destination path
                symlink_or_copy(source_paths[0], dest_path)
            else:
                # refuse to map multiple source files to one destination (undefined behaviour)
                raise TooManyFilesError(dest_path)
