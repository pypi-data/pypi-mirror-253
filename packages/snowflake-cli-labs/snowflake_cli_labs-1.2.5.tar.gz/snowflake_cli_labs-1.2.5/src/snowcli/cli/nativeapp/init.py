from __future__ import annotations

import logging
import os
from re import fullmatch
from pathlib import Path
from tempfile import TemporaryDirectory
from click.exceptions import ClickException
from shutil import move, rmtree
from strictyaml import load, as_document


from typing import Optional
from snowcli.cli.project.definition import DEFAULT_USERNAME
from snowcli.cli.project.util import clean_identifier, get_env_username
from snowcli.cli.common.utils import generic_render_template
from snowcli.cli.project.definition_manager import DefinitionManager


log = logging.getLogger(__name__)

SNOWFLAKELABS_GITHUB_URL = "https://github.com/Snowflake-Labs/native-apps-templates"
BASIC_TEMPLATE = "native-apps-basic"

# Based on first two rules for unquoted object identifier: https://docs.snowflake.com/en/sql-reference/identifiers-syntax
PROJECT_NAME_REGEX = r"(^[a-zA-Z_])([a-zA-Z0-9_$]{0,254})"


class InitError(ClickException):
    """
    Native app project could not be initiated due to an underlying error.
    """

    def __init__(self):
        super().__init__(self.__doc__)


class ProjectNameInvalidError(ClickException):
    """
    Intended project name does not qualify as a valid identifier.
    """

    def __init__(self):
        super().__init__(self.__doc__)


class RenderingFromJinjaError(ClickException):
    """
    Could not complete rendering file from Jinja template.
    """

    def __init__(self, name: str):
        super().__init__(
            f"Could not complete rendering file from Jinja template: {name}"
        )


class CannotInitializeAnExistingProjectError(ClickException):
    """
    Cannot initialize a new project within an existing Native Application project.
    """

    def __init__(self):
        super().__init__(self.__doc__)


class DirectoryAlreadyExistsError(ClickException):
    """
    Directory already contains a project with the intended name.
    """

    name: str

    def __init__(self, name: str):
        super().__init__(
            f"This directory already contains a sub-directory called {name}. Please try a different name."
        )
        self.name = name


def is_valid_project_name(project_name: str):
    return fullmatch(PROJECT_NAME_REGEX, project_name) is not None


def render_snowflake_yml(parent_to_snowflake_yml: Path):
    """
    Create a snowflake.yml file from a jinja template at a given path.

    Args:
        parent_to_snowflake_yml (Path): The parent directory of snowflake.yml.jinja, and later snowflake.yml

    Returns:
        None
    """

    snowflake_yml_jinja = "snowflake.yml.jinja"

    try:
        generic_render_template(
            template_path=parent_to_snowflake_yml / snowflake_yml_jinja,
            data={"project_name": parent_to_snowflake_yml.name},
            output_file_path=parent_to_snowflake_yml / "snowflake.yml",
        )
        os.remove(parent_to_snowflake_yml / snowflake_yml_jinja)
    except Exception as err:
        log.error(err)
        raise RenderingFromJinjaError(snowflake_yml_jinja)


def render_nativeapp_readme(parent_to_readme: Path, project_name: str):
    """
    Create a README.yml file from a jinja template at a given path.

    Args:
        parent_to_readme (Path): The parent directory of README.md.jinja, and later README.md

    Returns:
        None
    """

    readme_jinja = "README.md.jinja"

    default_application_name_prefix = clean_identifier(project_name)
    default_application_name_suffix = clean_identifier(
        get_env_username() or DEFAULT_USERNAME
    )

    try:
        generic_render_template(
            template_path=parent_to_readme / readme_jinja,
            data={
                "application_name": f"{default_application_name_prefix}_{default_application_name_suffix}"
            },
            output_file_path=parent_to_readme / "README.md",
        )
        os.remove(parent_to_readme / readme_jinja)
    except Exception as err:
        log.error(err)
        raise RenderingFromJinjaError(readme_jinja)


def replace_snowflake_yml_name_with_project(target_directory: Path):
    """
    Replace the native_app schema's "name" field in a snowflake.yml file with its parent directory name, i.e. the native app project, as the default start.
    This does not change the name in any other snowflake.*.yml as snowflake.yml is the base file and all others are overrides for the user to customize.

    Args:
        target_directory (str): The directory containing snowflake.yml at its root.

    Returns:
        None
    """

    path_to_snowflake_yml = target_directory / "snowflake.yml"
    contents = None

    with open(path_to_snowflake_yml) as f:
        contents = load(f.read()).data

    if "native_app" in contents and "name" in contents["native_app"]:
        contents["native_app"]["name"] = target_directory.name
        with open(path_to_snowflake_yml, "w") as f:
            f.write(as_document(contents).as_yaml())
    # If there are no such keys, the Definition Manager will catch that during validation


def validate_and_update_snowflake_yml(target_directory: Path):
    """
    Update the native_app name key in the snowflake.yml file and perform validation on the entire file.

    Args:
        target_directory (str): The directory containing snowflake.yml at its root.

    Returns:
        None
    """
    # 1. Determine if a snowflake.yml file exists, at the very least
    definition_manager = DefinitionManager(target_directory)

    # 2. Change the project name in snowflake.yml to project_name
    replace_snowflake_yml_name_with_project(target_directory=target_directory)

    # 3. Validate the Project Definition File(s)
    # We do not need to use the result of the call below, we just want to validate the file
    definition_manager.project_definition


def _init_with_url_and_no_template(
    current_working_directory: Path,
    project_name: str,
    git_url: str,
):
    """
    Initialize a Native Apps project with a git url but without any specific template specified by the user.

    Args:
        current_working_directory (str): The current working directory of the user where the project will be added.
        project_name (str): Name of the project to be created.
        git_url (str): The git URL to perform a clone from.

    Returns:
        None
    """

    target_directory: Optional[Path] = None
    try:
        # with TemporaryDirectory(dir=current_working_directory) as temp_dir:
        target_directory = current_working_directory / project_name
        target_directory.mkdir(parents=True, exist_ok=False)

        # Clone the repository with options.
        from git import Repo

        Repo.clone_from(
            url=git_url,
            to_path=target_directory,
            filter=["tree:0"],
            depth=1,
        )

        # Remove all git history
        rmtree(target_directory.joinpath(".git").resolve())

        # Non-Snowflake git URLs should not have jinja files in their directory structure.
        # If they do, snowCLI is not responsible for rendering them during init.
        # If the SNOWFLAKELABS_GITHUB_URL is provided here, rendering is additionally skipped as there is no jinja file at the root.

        # If not an official Snowflake Native App template
        if git_url != SNOWFLAKELABS_GITHUB_URL:
            validate_and_update_snowflake_yml(target_directory=target_directory)

    except Exception as err:
        # If there was any error, validation on Project Definition file or otherwise,
        # there should not be any Native Apps Project left after this.
        if target_directory:
            rmtree(target_directory.resolve())

        log.error(err)
        raise InitError()


def _init_with_url_and_template(
    current_working_directory: Path, project_name: str, git_url: str, template: str
):
    """
    Initialize a Native Apps project with a git URL and a specific template within the git URL.

    Args:
        current_working_directory (str): The current working directory of the user where the project will be added.
        project_name (str): Name of the project to be created.
        git_url (str): The git URL to perform a clone from.
        template (str): A template within the git URL to use, all other directories and files outside the template will be discarded.

    Returns:
        None
    """

    path_to_project: Optional[Path] = None
    try:
        with TemporaryDirectory(dir=current_working_directory) as temp_dir:
            # Clone the repository in the temporary directory with options.
            from git import Repo

            Repo.clone_from(
                url=git_url,
                to_path=temp_dir,
                filter=["tree:0"],
                depth=1,
            )

            # Move native-apps-basic to current_working_directory and rename to name
            move(
                src=current_working_directory / temp_dir / template,
                dst=current_working_directory / project_name,
            )

        path_to_project = current_working_directory / project_name
        # Rendering should be conditinal on the below combination,
        # as right now, we only allow rendering of a jinja file within this combination
        if (git_url == SNOWFLAKELABS_GITHUB_URL) and (template == BASIC_TEMPLATE):
            # Render snowflake.yml file from its jinja template
            render_snowflake_yml(parent_to_snowflake_yml=path_to_project)

            # Render README.md file from its jinja template
            render_nativeapp_readme(
                parent_to_readme=path_to_project / "app",
                project_name=project_name,
            )

        # If not an official Snowflake Native App template
        if git_url != SNOWFLAKELABS_GITHUB_URL:
            validate_and_update_snowflake_yml(target_directory=path_to_project)

    except Exception as err:
        # If there was any error, validation on Project Definition file or otherwise,
        # there should not be any Native Apps Project left after this.
        if path_to_project:
            rmtree(path_to_project.resolve())

        log.error(err)
        raise InitError()


def nativeapp_init(
    name: str, git_url: Optional[str] = None, template: Optional[str] = None
):
    """
    Initialize a Native Apps project in the user's current working directory, with or without the use of a template.

    Args:
        name (str): Name of the project to be created.
        git_url (str): The git URL to perform a clone from.
        template (str): A template within the git URL to use, all other directories and files outside the template will be discarded.

    Returns:
        None
    """

    current_working_directory = Path.cwd()

    # If the intended project name is not a valid identifier, faili init command
    if not is_valid_project_name(name):
        raise ProjectNameInvalidError()

    # If current directory is already contains a file named snowflake.yml, i.e. is a native apps project, fail init command.
    # We do not validate contents of the yml here though.
    path_to_snowflake_yml = current_working_directory / "snowflake.yml"
    if path_to_snowflake_yml.is_file():
        raise CannotInitializeAnExistingProjectError()

    # If a subdirectory with the same name as name exists in the current directory, fail init command
    path_to_project = current_working_directory / name
    if path_to_project.exists():
        raise DirectoryAlreadyExistsError(name)

    if (
        git_url and not template
    ):  # If user provided a git url but no template, use the full clone
        _init_with_url_and_no_template(
            current_working_directory=current_working_directory,
            project_name=name,
            git_url=git_url,
        )
    else:  # If user provided some other combination of git url and template, only prioritize the template cloning
        _init_with_url_and_template(
            current_working_directory=current_working_directory,
            project_name=name,
            git_url=git_url if git_url else SNOWFLAKELABS_GITHUB_URL,
            template=template if template else BASIC_TEMPLATE,
        )
