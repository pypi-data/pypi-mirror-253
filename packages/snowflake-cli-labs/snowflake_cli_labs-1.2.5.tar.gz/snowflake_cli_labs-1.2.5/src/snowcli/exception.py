from click.exceptions import ClickException


class EnvironmentVariableNotFoundError(ClickException):
    def __init__(self, env_variable_name: str):
        super().__init__(f"Environment variable {env_variable_name} not found")


class MissingConfiguration(ClickException):
    pass


class InvalidConnectionConfiguration(ClickException):
    def format_message(self):
        return f"Invalid connection configuration. {self.message}"


class SnowflakeConnectionError(ClickException):
    def __init__(self, snowflake_err: Exception):
        super().__init__(f"Could not connect to Snowflake. Reason: {snowflake_err}")


class UnsupportedConfigSectionTypeError(Exception):
    def __init__(self, section_type: type):
        super().__init__(f"Unsupported configuration section type {section_type}")


class OutputDataTypeError(ClickException):
    def __init__(self, got_type: type, expected_type: type):
        super().__init__(f"Got {got_type} type but expected {expected_type}")


class CommandReturnTypeError(ClickException):
    def __init__(self, got_type: type):
        super().__init__(f"Commads have to return OutputData type, but got {got_type}")
