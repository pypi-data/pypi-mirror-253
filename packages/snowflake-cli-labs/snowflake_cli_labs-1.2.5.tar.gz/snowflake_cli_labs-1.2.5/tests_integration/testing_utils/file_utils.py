def replace_text_in_file(file_path: str, to_replace: str, replacement: str) -> None:
    with open(file_path, "r") as file:
        text = file.read()
    text = text.replace(to_replace, replacement)
    with open(file_path, "w") as file:
        file.write(text)
