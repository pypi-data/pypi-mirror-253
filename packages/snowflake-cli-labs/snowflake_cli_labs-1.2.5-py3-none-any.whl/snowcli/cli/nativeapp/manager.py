from __future__ import annotations

from pathlib import Path
from functools import cached_property
from typing import List, Optional, Callable

from snowcli.cli.common.sql_execution import SqlExecutionMixin
from snowflake.connector import DictCursor

from .artifacts import build_bundle, translate_artifact, ArtifactMapping
from ..project.definition_manager import DefinitionManager


def find_row(cursor: DictCursor, predicate: Callable[[dict], bool]) -> Optional[dict]:
    """Returns the first row that matches the predicate, or None."""
    return next(
        (row for row in cursor.fetchall() if predicate(row)),
        None,
    )


class NativeAppManager(SqlExecutionMixin):
    definition_manager: DefinitionManager

    def __init__(self, search_path: Optional[str] = None):
        super().__init__()
        self.definition_manager = DefinitionManager(search_path or "")

    @property
    def project_root(self) -> Path:
        return self.definition_manager.project_root

    @property
    def definition(self) -> dict:
        return self.definition_manager.project_definition["native_app"]

    @cached_property
    def artifacts(self) -> List[ArtifactMapping]:
        return [translate_artifact(item) for item in self.definition["artifacts"]]

    @cached_property
    def deploy_root(self) -> Path:
        return Path(self.project_root, self.definition["deploy_root"])

    def build_bundle(self) -> None:
        build_bundle(self.project_root, self.deploy_root, self.artifacts)
