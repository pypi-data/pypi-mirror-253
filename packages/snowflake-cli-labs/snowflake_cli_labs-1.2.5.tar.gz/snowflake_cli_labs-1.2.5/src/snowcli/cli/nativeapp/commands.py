from typing import Optional

import logging
import typer

from snowcli.cli.common.decorators import global_options
from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS
from snowcli.output.decorators import with_output

from .init import nativeapp_init
from .manager import NativeAppManager

from snowcli.output.types import (
    CommandResult,
    MessageResult,
)

app = typer.Typer(
    context_settings=DEFAULT_CONTEXT_SETTINGS,
    hidden=True,
    name="app",
    help="Manage Native Apps in Snowflake",
)

log = logging.getLogger(__name__)

ProjectArgument = typer.Option(
    None,
    "-p",
    "--project",
    help="Path where the Native Apps project resides. Defaults to current working directory",
    show_default=False,
)


@app.command("init")
@with_output
@global_options
def app_init(
    name: str = typer.Argument(
        ..., help="Name of the Native Apps project to be initiated."
    ),
    git_url: str = typer.Option(
        None,
        help="A git URL to use as template for the Native Apps project. Example: https://github.com/Snowflake-Labs/native-apps-templates.git for all official Snowflake templates.",
    ),
    template: str = typer.Option(
        None,
        help="A specific directory within the git URL to use as template for the Native Apps project. Example: Default is native-app-basic if --git-url is https://github.com/Snowflake-Labs/native-apps-templates.git, and None if any other --git-url.",
    ),
    **options,
) -> CommandResult:
    """
    Initialize a Native Apps project, optionally with a --git-url and a --template.
    """
    nativeapp_init(name, git_url, template)
    return MessageResult(
        f"Native Apps project {name} has been created in your local directory."
    )


@app.command("bundle", hidden=True)
@with_output
@global_options
def app_bundle(
    project_path: Optional[str] = ProjectArgument,
    **options,
) -> CommandResult:
    """
    Prepares a local folder with configured app artifacts.
    """
    manager = NativeAppManager(project_path)
    manager.build_bundle()
    return MessageResult(f"Bundle generated at {manager.deploy_root}")
