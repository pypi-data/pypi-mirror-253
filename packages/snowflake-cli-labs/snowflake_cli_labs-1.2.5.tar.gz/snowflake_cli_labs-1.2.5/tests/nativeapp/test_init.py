from snowcli.cli.nativeapp.init import (
    is_valid_project_name,
    render_snowflake_yml,
    render_nativeapp_readme,
    nativeapp_init,
    replace_snowflake_yml_name_with_project,
    validate_and_update_snowflake_yml,
    _init_with_url_and_no_template,
    _init_with_url_and_template,
    CannotInitializeAnExistingProjectError,
    DirectoryAlreadyExistsError,
    InitError,
    ProjectNameInvalidError,
)
from snowcli.exception import MissingConfiguration
from tests.testing_utils.fixtures import *
from textwrap import dedent
from secrets import choice
from string import ascii_letters, digits

PROJECT_NAME = "demo_na_project"
MAX_ALLOWED_NUM_CHARACTERS = 255

# --------------------------------------
# ----- Tests for Helper Methods -------
# --------------------------------------


@pytest.mark.parametrize(
    "project_name, expected",
    [
        ("_", True),  # Edge Case: Only One Character from the Allowed List
        ("A", True),  # Edge Case: Only One Character from the Allowed List
        ("9", False),  # Edge Case: Only One Character not from the Allowed List
        ("_aB3_$", True),  # Test all allowed character types
        ("__", True),  # Test all allowed character types
        (
            "".join(choice(ascii_letters) for i in range(MAX_ALLOWED_NUM_CHARACTERS)),
            True,
        ),
        (
            "".join(
                choice(ascii_letters) for i in range(MAX_ALLOWED_NUM_CHARACTERS - 2)
            ).join("*%"),
            False,
        ),
    ],
)
def test_is_valid_project_name(project_name, expected):
    assert is_valid_project_name(project_name) == expected


def test_render_snowflake_yml(other_directory):
    temp_dir = Path(other_directory)
    create_named_file(
        file_name="snowflake.yml.jinja",
        dir=temp_dir,
        contents=[
            dedent(
                """\
            native_app:
                name: {{project_name}}
                artifacts:
                    - app/setup_script.sql
            """
            )
        ],
    )
    expected = dedent(
        f"""\
        native_app:
            name: {temp_dir.name}
            artifacts:
                - app/setup_script.sql

        """
    )
    render_snowflake_yml(temp_dir)
    assert Path.exists(temp_dir / "snowflake.yml")
    assert not Path.exists(temp_dir / "snowflake.yml.jinja")
    assert temp_dir.joinpath("snowflake.yml").read_text() == expected


@mock.patch("os.getenv", return_value="pytest_user")
def test_render_nativeapp_readme(mock_get_env_username, other_directory):
    temp_dir = Path(other_directory)
    create_named_file(
        file_name="README.md.jinja",
        dir=temp_dir,
        contents=[
            dedent(
                """\
            ### Calling a function
            SELECT {{application_name}}.versioned_schema.hello_world();
            which should output 'hello world!'
            ```
            SELECT {{application_name}}.versioned_schema.hello_world();
            ```
            """
            )
        ],
    )
    expected = dedent(
        """\
            ### Calling a function
            SELECT random_project_pytest_user.versioned_schema.hello_world();
            which should output 'hello world!'
            ```
            SELECT random_project_pytest_user.versioned_schema.hello_world();
            ```
            
            """
    )
    render_nativeapp_readme(temp_dir, "random_project")
    assert Path.exists(temp_dir / "README.md")
    assert not Path.exists(temp_dir / "README.md.jinja")
    assert temp_dir.joinpath("README.md").read_text() == expected


def test_replace_snowflake_yml_name_with_project_populated_file(other_directory):
    temp_dir = Path(other_directory)
    create_named_file(
        file_name="snowflake.yml",
        dir=temp_dir,
        contents=[
            dedent(
                """\
            native_app:
                name: old_value
                artifacts:
                    - app/setup_script.sql
            """
            )
        ],
    )
    expected = dedent(
        f"""\
        native_app:
          name: {temp_dir.name}
          artifacts:
          - app/setup_script.sql
        """
    )
    path_to_snowflake_yml = temp_dir / "snowflake.yml"
    replace_snowflake_yml_name_with_project(temp_dir)
    assert Path.exists(path_to_snowflake_yml)
    assert path_to_snowflake_yml.read_text() == expected


def test_replace_snowflake_yml_name_with_project_empty_file(other_directory):
    temp_dir = Path(other_directory)
    create_named_file(
        file_name="snowflake.yml",
        dir=temp_dir,
        contents=[""],
    )
    expected = dedent(
        f"""\
        native_app:
          name: {temp_dir.name}

        """
    )
    path_to_snowflake_yml = temp_dir / "snowflake.yml"
    replace_snowflake_yml_name_with_project(temp_dir)
    assert Path.exists(path_to_snowflake_yml)
    assert not path_to_snowflake_yml.read_text().strip()


def test_validate_and_update_snowflake_yml_w_missing_yml(other_directory):
    temp_dir = Path(other_directory)

    with pytest.raises(MissingConfiguration):
        validate_and_update_snowflake_yml(target_directory=temp_dir)


# ------------------------------------------------------
# ----- Tests for _init_with_url_and_no_template -------
# ------------------------------------------------------


def test_init_with_url_and_no_template_fail_on_validation(temp_dir):
    # Should fail as git URL provided has no snowflake.yml file to validate
    with pytest.raises(InitError):
        _init_with_url_and_no_template(
            current_working_directory=Path.cwd(),
            project_name="new_project",
            git_url="https://github.com/Snowflake-Labs/sf-samples",
        )
    assert not Path.exists(Path.cwd() / "new_project")


@mock.patch("git.Repo.clone_from", side_effect=None)
def test_init_with_url_and_no_template_w_native_app_url(mock_clone_from, temp_dir):
    # Prepare fake repository
    fake_repo = "fake_repo"
    current_working_directory = Path.cwd()
    current_working_directory.joinpath(fake_repo, ".git").mkdir(
        parents=True, exist_ok=False
    )
    create_named_file(
        file_name="snowflake.yml",
        dir=current_working_directory / fake_repo,
        contents=[
            dedent(
                """\
            definition_version: 1
            native_app:
                name: minimal

                artifacts:
                    - setup.sql
                    - README.md
            """
            )
        ],
    )

    # Now mock the mkdir call so that a new project is not created, and the fake repo above is used instead
    with mock.patch("pathlib.Path.mkdir") as mock_mkdir:
        mock_mkdir.return_value = None
        mock_mkdir.side_effect = None

        _init_with_url_and_no_template(
            current_working_directory=Path.cwd(),
            project_name=fake_repo,
            git_url="https://github.com/Snowflake-Labs/native-apps-templates",
        )

        dest = Path.cwd() / fake_repo
        assert dest.exists()
        assert not Path.exists(dest / ".git")


@mock.patch("git.Repo.clone_from", side_effect=None)
def test_init_with_url_and_no_template_w_random_url(mock_clone_from, temp_dir):

    # Prepare fake repository
    fake_repo = "fake_repo"
    current_working_directory = Path.cwd()
    current_working_directory.joinpath(fake_repo, ".git").mkdir(
        parents=True, exist_ok=False
    )
    create_named_file(
        file_name="snowflake.yml",
        dir=current_working_directory / fake_repo,
        contents=[
            dedent(
                """\
            definition_version: 1
            native_app:
                name: minimal

                artifacts:
                    - setup.sql
                    - README.md
            """
            )
        ],
    )

    expected = dedent(
        f"""\
        definition_version: 1
        native_app:
          name: fake_repo
          artifacts:
          - setup.sql
          - README.md
        """
    )

    # Now mock the mkdir call so that a new project is not created, and the fake repo above is used instead
    with mock.patch("pathlib.Path.mkdir") as mock_mkdir:
        mock_mkdir.return_value = None
        mock_mkdir.side_effect = None

        _init_with_url_and_no_template(
            current_working_directory=Path.cwd(),
            project_name=fake_repo,
            git_url="https://github.com/Snowflake-Labs/sf-samples",
        )

        dest = Path.cwd() / fake_repo
        assert dest.exists()
        assert not Path.exists(dest / ".git")
        assert dest.joinpath("snowflake.yml").read_text() == expected


# ---------------------------------------------------
# ----- Tests for _init_with_url_and_template -------
# ---------------------------------------------------


@mock.patch("git.Repo.clone_from", side_effect=None)
@mock.patch("snowcli.cli.nativeapp.init.move", side_effect=None)
@mock.patch("os.getenv", return_value="pytest_user")
def test_init_with_url_and_template_w_native_app_url_and_template(
    mock_clone_from, mock_move, mock_getenv, temp_dir
):
    # Prepare fake repository
    fake_repo = "fake_repo"
    current_working_directory = Path.cwd()
    current_working_directory.joinpath(fake_repo, "app").mkdir(
        parents=True, exist_ok=False
    )
    create_named_file(
        file_name="snowflake.yml.jinja",
        dir=current_working_directory / fake_repo,
        contents=[
            dedent(
                """\
            definition_version: 1
            native_app:
                name: {{project_name}}

                artifacts:
                    - setup.sql
                    - README.md
            """
            )
        ],
    )

    expected_snowflake_yml = dedent(
        """\
            definition_version: 1
            native_app:
                name: fake_repo
            
                artifacts:
                    - setup.sql
                    - README.md
            
            """
    )

    create_named_file(
        file_name="README.md.jinja",
        dir=current_working_directory / fake_repo / "app",
        contents=[dedent("{{application_name}}")],
    )
    expected_readme = dedent("fake_repo_pytest_user\n")

    _init_with_url_and_template(
        current_working_directory=Path.cwd(),
        project_name=fake_repo,
        git_url="https://github.com/Snowflake-Labs/native-apps-templates",
        template="native-apps-basic",
    )

    fake_repo_path = current_working_directory / fake_repo
    assert fake_repo_path.exists()
    assert not Path.exists(fake_repo_path / ".git")
    assert (
        fake_repo_path.joinpath("snowflake.yml").read_text() == expected_snowflake_yml
    )
    assert fake_repo_path.joinpath("app", "README.md").read_text() == expected_readme


@mock.patch("git.Repo.clone_from", side_effect=None)
@mock.patch("snowcli.cli.nativeapp.init.move", side_effect=None)
@mock.patch("os.getenv", return_value="pytest_user")
def test_init_with_url_and_template_w_random_url_and_template(
    mock_clone_from, mock_move, mock_getenv, temp_dir, snapshot
):
    # Prepare fake repository
    fake_repo = "fake_repo"
    current_working_directory = Path.cwd()
    current_working_directory.joinpath(fake_repo).mkdir(parents=True, exist_ok=False)
    create_named_file(
        file_name="snowflake.yml",
        dir=current_working_directory / fake_repo,
        contents=[
            dedent(
                """\
            definition_version: 1
            native_app:
                name: <old_name>

                artifacts:
                    - setup.sql
                    - README.md
            """
            )
        ],
    )

    _init_with_url_and_template(
        current_working_directory=Path.cwd(),
        project_name=fake_repo,
        git_url="https://github.com/Snowflake-Labs/sf-samples",
        template="samples",
    )

    fake_repo_path = current_working_directory / fake_repo
    assert fake_repo_path.exists()
    assert not Path.exists(fake_repo_path / ".git")
    assert fake_repo_path.joinpath("snowflake.yml").read_text() == snapshot


# --------------------------------------
# ----- Tests for Short Circuits -------
# --------------------------------------


@mock.patch("pathlib.Path.is_file", return_value=True)
def test_init_w_existing_yml(mock_path_is_file):
    with pytest.raises(CannotInitializeAnExistingProjectError):
        nativeapp_init(name=PROJECT_NAME)


@mock.patch("pathlib.Path.exists", return_value=True)
def test_init_w_existing_directory(mock_path_exists):
    with pytest.raises(DirectoryAlreadyExistsError):
        nativeapp_init(name=PROJECT_NAME)


@mock.patch("snowcli.cli.nativeapp.init.fullmatch", return_value=None)
def test_init_w_invalid_project_name(mock_fullmatch):
    with pytest.raises(ProjectNameInvalidError):
        nativeapp_init(PROJECT_NAME)
