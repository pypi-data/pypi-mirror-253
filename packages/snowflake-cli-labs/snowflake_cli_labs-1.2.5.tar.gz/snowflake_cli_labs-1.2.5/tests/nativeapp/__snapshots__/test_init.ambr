# serializer version: 1
# name: test_init_with_url_and_template_w_random_url_and_template
  '''
  definition_version: 1
  native_app:
    name: fake_repo
    artifacts:
    - setup.sql
    - README.md
  
  '''
# ---
